README.org

```diff
@@ -233,6 +233,25 @@ This is great for performance, but means the data can become stale if you modify
 The =bibtex-actions-refresh= command will reload the caches, and you can call this manually.
 You can also call any of the =bibtex-actions= commands with a prefix argument: =C-u M-x bibtex-actions-insert-key=.
 
+Although not default =Bibtex-actions= also provides convenience functions for auto-refreshing cache when bib files change using filenotify. There are two such functions. =bibtex-actions-with-filenotify-local= and =bibtex-actions-filenotify-global=. The following snippet will refresh caches when one of the bib files change
+
+#+BEGIN_SRC emacs-lisp
+(bibtex-actions-with-filenotify-global #'bibtex-actions-refresh)
+
+(add-hook 'LaTeX-mode-hook (lambda () (bibtex-actions-with-filenotify-local #'bibtex-actions-refresh)))
+(add-hook 'org-mode-hook (lambda () (bibtex-actions-with-filenotify-local #'bibtex-actions-refresh)))
```

Comment:
Question: do we need to do anything to remove the inotify watches, or does the fact that this is a hook take care of that?

Reply:
The function `bibtex-actions-with-filenotify` takes care of removing the watches. It adds a buffer local hook to `kill-buffer-hook` to remove these watches when the buffer is killed.

Reply:
I should probably clarify `bibtex-actions-with-filenotify-local` takes care of it. The global version just returns the descriptors. The global watches need to be killed manually.

Reply:
> I should probably clarify `bibtex-actions-with-filenotify-local` takes care of it. The global version just returns the descriptors. The global watches need to be killed manually.

Is that a necessary limitation?

Reply:
No, I just don't know when the global watches should be removed.

Reply:
I think bibtex-completion creates and destroys them per session?

https://github.com/tmalsburg/helm-bibtex/blob/12079bb09f203dda5cc2dd003bd60a6ad490f762/bibtex-completion.el#L396

Reply:
BTW, one of the reasons I bypassed the bibtex-completion support for this is I wanted more flexibility, and wasn't sure how best to do that.

For example, I want to have a watch on my notes directory.

Reply:
> I think bibtex-completion creates and destroys them per session?
> 
> https://github.com/tmalsburg/helm-bibtex/blob/12079bb09f203dda5cc2dd003bd60a6ad490f762/bibtex-completion.el#L396

I think this is because the design of bibtex-completions is a bit adhoc. In particular since local bib is used through a let binding the collection of bib files can change from call to call. This is not the case for us. 

Let us assume for a moment that the collection doesn't changes. In that case what we have is correct. The watches on the global bin should exist unless manually killed and those on local files should be killed with the buffer. Though there is a problem with the code as is. If the major mode hook is run multiple times which can happen there will be multiple watches on the file. I have fixed this and will do another pr.

Now let us consider the case when a bib file is added and removed which I think should be rare. For the local bib files this isn't handled by bibtex-completions. To detect this a reftex scan is needed. I think the best solution here is to have a manual refresh function that runs the scan and adjusts the watches. A user can run then call it after adding/removing bib files.

Reply:
> I think the best solution here is to have a manual refresh function that runs the scan and adjusts the watches. A user can run then call it after adding/removing bib files

I agree.

But per my earlier comment, I would allow it to be configured to include note and library paths too.

Reply:
> But per my earlier comment, I would allow it to be configured to include note and library paths too.

Ok, so basically add an optional list parameter that contains any extra files to be watched?

Reply:
> so basically add an optional list parameter that contains any extra files to be watched?

Yes, but where the paths can be specific via variable.

E.g. if I have the notes-path variable set, I should be able to tell it to use that.

Reply:
> E.g. if I have the notes-path variable set, I should be able to tell it to use that.

You can pass the variable as the optional argument to the function. Or are you thinking of something else?

Reply:
Only that it's easy and flexible for the user.

Reply:
I created the pr so we can discuss there if something needs changing.